baseband/guppi/header.py

```diff
@@ -0,0 +1,379 @@
+        keywords have been processed.
+
+        Parameters
+        ----------
+        verify : bool, optional
+            If `True` (default), verify integrity after updating.
+        **kwargs
+            Arguments used to set keywords and properties.
+        """
+        verify = kwargs.pop('verify', True)
+        # Remove kwargs that set properties, in correct order.
+        extras = [(key, kwargs.pop(key)) for key in self._properties
+                  if key in kwargs]
+        # Update the normal keywords.
+        super(GUPPIHeader, self).update(kwargs)
+        # Now set the properties.
+        for attr, value in extras:
+            setattr(self, attr, value)
+        if verify:
+            self.verify()
+
+    def __setitem__(self, key, value):
+        if not self.mutable:
+            raise TypeError("immutable {0} does not support assignment."
+                            .format(type(self).__name__))
+        if isinstance(value, tuple):
+            value, comment = value
+            self.comments[key.upper()] = comment
+
+        super(GUPPIHeader, self).__setitem__(key.upper(), value)
+
+    @property
+    def nbytes(self):
+        """Size of the header in bytes."""
+        return (len(self) + 1) * 80
+
+    @property
+    def payload_nbytes(self):
+        """Size of the payload in bytes."""
+        return self['BLOCSIZE']
+
+    @payload_nbytes.setter
+    def payload_nbytes(self, payloadsize):
+        self['BLOCSIZE'] = payloadsize
+
+    @property
+    def frame_nbytes(self):
+        """Size of the frame in bytes."""
+        return self.nbytes + self.payload_nbytes
+
+    @frame_nbytes.setter
+    def frame_nbytes(self, frame_nbytes):
+        self.payload_nbytes = frame_nbytes - self.nbytes
+
+    @property
+    def bps(self):
+        """Bits per elementary sample."""
+        return self['NBITS']
+
+    @bps.setter
+    def bps(self, bps):
+        self['NBITS'] = bps
+
+    @property
+    def complex_data(self):
+        """Whether the data are complex."""
+        return self['OBSNCHAN'] != 1
+
+    @property
+    def npol(self):
+        """Number of polarisations."""
+        return self['NPOL'] // (2 if self.complex_data else 1)
+
+    @npol.setter
+    def npol(self, npol):
+        self['NPOL'] = npol * (2 if self.complex_data else 1)
+
+    @property
+    def nchan(self):
+        """Number of channels."""
+        return self['OBSNCHAN']
+
+    @nchan.setter
+    def nchan(self, nchan):
+        self['OBSNCHAN'] = nchan.__index__()
+
+    @property
+    def sample_shape(self):
+        """Shape of a sample in the payload (npol, nchan)."""
+        return self.npol, self.nchan
+
+    @sample_shape.setter
+    def sample_shape(self, sample_shape):
+        # Need to set nchan first to properly set npol, since nchan is
+        # connected to complex_data.
+        self.nchan = sample_shape[1]
+        self.npol = sample_shape[0]
+
+    @property
```

Comment:
I wonder about having this method just for this class. It is equivalent to `sample_size * bps * (2 if self.complex else 1)` and valid in principle for all classes. My tendency would be to either not have it or keep it private. It is also unusual in not having a setter.

Reply:
Let's make it private for now, then consider making a general `bpfs` property if it's ever needed.  This also lets me shorten it to `_bpfs` making some of the lines less unwieldy.

Obviously giving this a setter makes no sense, but since `npol`, etc. have setters I don't think it's a good idea to make this a lazy property.  I also don't think it's that unusual - `nbytes` doesn't have a setter either.

Reply:
OK